### DISCUSSION TITLE
发送hex数据加上modbus16校验码

### DISCUSSION THREAD

**Discussion by sky-shine** [chenxuuu/llcom#136] [脚本 Scripts publish]
- sky-shine
- 发hex加上modbus16校验码
- 发送hex数据加上modbus16校验码，在处理发送数据的脚本里设置，发送hex数据会在后面加上modbusCRC校验，如果需要其他16位校验码，改XOR应该就可以，这里的0xA001是0x8005的按位反转
- 0
- 在处理发送数据的脚本里设置，发送hex数据会在后面加上modbusCRC校验，如果需要其他16位校验码，改XOR应该就可以，这里的0xA001是0x8005的按位反转

```lua
function modbuscrc16(str)
	local flag=0xFFFF
	local temp=0xFFFF
	for i=1, string.len(str) do
	    temp=temp~string.byte(str,i)
	    for j=1, 8 do
	    	flag=temp&0x0001
	        temp=temp>>1
	        if (flag>0)
	        --if LSB is set, XOR is 0xA001
	        then temp =temp ~ 0xA001
	        end
	    end
	    
	end
	--高低字节顺序对换
	temp = ((temp << 8)&0xFF00) | (temp >> 8);
	--返回4字节hex string
	return string.format("%04X",temp)
end

return uartData..modbuscrc16(uartData):fromHex()
```

**Comment by 4311465** [comment]
我运行就报错19:50:16:6639]lua代码报错了：
XLua.LuaException: D:\github\llcom\llcom\bin\x64\user_script_run\test1.lua:4: bad argument #1 to 'len' (string expected, got nil)
stack traceback:
	[C]: in function 'string.len'
